File: Leetcode/Python3/1254. Number of Closed Islands/solution.py
# https://leetcode.com/problems/number-of-closed-islands/

from collections import deque


dx = [0, 1, 0, -1]
dy = [1, 0, -1, 0]
class Solution:
    def closedIsland(self, grid: List[List[int]]) -> int:
        def bfs(x, y):
            close = True
            q = deque()
            q.append((x, y))
            grid[x][y] = 1
            while q:
                x, y = q.popleft()
                # If the island is connected to the wall of the grid, it is not close.
                if x == 0 or x == m - 1 or y == 0 or y == n - 1:    
                            close = False
                for i in range(4):
                    nx = x + dx[i]
                    ny = y + dy[i]
                    if 0 <= nx < m and 0 <= ny < n and grid[nx][ny] == 0:
                        q.append((nx, ny))
                        grid[nx][ny] = 1
            return close
        m, n = len(grid), len(grid[0])
        answer = 0
        for idx_x in range(m):
            for idx_y in range(n):
                if grid[idx_x][idx_y] == 0:
                    answer += bfs(idx_x, idx_y) # True == 1, False == 0
        return answer

File: Leetcode/Typescript/1254. Number of Closed Islands/solution.ts
// https://leetcode.com/problems/number-of-closed-islands/description/

function closedIsland(grid: number[][]): number {
  let closed = 0;

  for (let y = 0; y < grid.length; y++) {
    for (let x = 0; x < grid[0].length; x++) {
      if (grid[y][x] === 0) {
        closed += checkClosed(grid, x, y);
      }
    }
  }

  return closed;
}

function checkClosed(grid: number[][], x: number, y: number): number {
  const stack: number[][] = [[x, y]];
  const dx = [0, 0, 1, -1],
    dy = [1, -1, 0, 0];
  let isClosed = true;

  while (stack.length > 0) {
    const [cur_x, cur_y] = stack.pop();

    for (let i = 0; i < 4; i++) {
      const new_x = cur_x + dx[i],
        new_y = cur_y + dy[i];
      if (
        new_x < 0 ||
        grid[0].length <= new_x ||
        new_y < 0 ||
        grid.length <= new_y
      ) {
        isClosed = false;
        continue;
      }
      if (grid[new_y][new_x] === 0) {
        stack.push([new_x, new_y]);
        grid[new_y][new_x] = 1;
      }
    }
  }

  return isClosed ? 1 : 0;
}
